### PR TITLE
ISPN-12177 Persist IRAC version during shutdown

### PR DESCRIPTION
**Pull request by pruivo**
https://issues.redhat.com/browse/ISPN-12177

### CODE REVIEW

**Comment by tristantarrant** [comment]
Conflicts

**Review by wburns — CHANGES_REQUESTED** [review]
So I see where it loads from the file, but isn't there supposed to be something to read the state from the other site? Or is that handled elsewhere?

And what about if we have data in a store, guessing that is a different PR?

File: core/src/main/java/org/infinispan/commands/irac/IracMetadataRequestCommand.java [hunk 1]
```diff
@@ -90,6 +92,7 @@ public String toString() {
             "cacheName=" + cacheName +
             ", segment=" + segment +
             ", topologyId=" + topologyId +
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/commands/irac/IracMetadataRequestCommand.java:94]: not really. It makes sure the node has at least that topology installed before handling the executing the command.

[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/commands/irac/IracMetadataRequestCommand.java:94]: Shouldn't it be part of the `readFrom` and `writeTo` though?

File: core/src/main/java/org/infinispan/container/versioning/irac/TopologyIracVersion.java [hunk 2]
```diff
@@ -36,6 +49,16 @@ public static TopologyIracVersion max(TopologyIracVersion v1, TopologyIracVersio
       return v1.compareTo(v2) < 0 ? v2 : v1;
    }
 
+   public static TopologyIracVersion fromString(String s) {
+      Matcher m = Pattern.compile(REGEX).matcher(s);
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/container/versioning/irac/TopologyIracVersion.java:53]: right!

File: core/src/main/java/org/infinispan/container/versioning/irac/TopologyIracVersion.java [hunk 3]
```diff
@@ -63,7 +86,7 @@ public int compareTo(TopologyIracVersion other) {
 
    @Override
    public String toString() {
-      return '(' + Integer.toString(topologyId) + ':' + version + ')';
+      return String.format(FMT, topologyId, version);
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/container/versioning/irac/TopologyIracVersion.java:89]: I've spent too much time in python xD

**Comment by pruivo** [comment]
Each node, when a rebalance happends, broadcast its version state to all other nodes (see `org.infinispan.container.versioning.irac.DefaultIracVersionGenerator#onTopologyChange()`)

**Comment by pruivo** [comment]
About the cache store, it uses the version in the context to generate a new higher version. Example, for non-tx caches: https://github.com/infinispan/infinispan/pull/8680/files#diff-8f85a96763cdc57d2a36e8a8b873178ab509a3083766c6d2ff9bb56a07a1a2c6R205

this covers "all" scenarios except if you use SKIP_CACHE_LOAD + !preload  + !conditional_write

**Comment by pruivo** [comment]
updated & rebased

File: core/src/main/java/org/infinispan/commands/irac/IracMetadataRequestCommand.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/commands/irac/IracMetadataRequestCommand.java:94]: no, the previous code was incorrect.
The `CacheRpcCommandExternalizer` [1] invokes the `ReplicableCommandExternalizer` [2] and the latter sends the topology id.

[1] https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/marshall/exts/CacheRpcCommandExternalizer.java#L142
[2] https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/marshall/exts/ReplicableCommandExternalizer.java#L82

**Comment by pruivo** [comment]
fixed test failure

**Comment by wburns** [comment]
Integrated into master, thanks @pruivo !

**Comment by pruivo** [comment]
yey!  thanks @wburns !

**Pull request merged**